File: backend/services/weather_service.py
# OPEN-METEO API
# No API key is required for this prototype.
# Weather data is fetched using latitude and longitude.

import urllib.request
import json
from typing import Dict, Any, Optional

class WeatherService:
    """
    Weather & Precipitation telemetry service.
    Fetches real-time observations and forecasts from Open-Meteo Forecast API.
    """
    
    OPEN_METEO_BASE_URL = "https://api.open-meteo.com/v1/forecast"

    @staticmethod
    def get_intensity_label(rate_mm_hr: Optional[float]) -> str:
        if rate_mm_hr is None:
            return "Unavailable"
        if rate_mm_hr >= 100:
            return "Cloudburst / Torrential"
        elif rate_mm_hr >= 50:
            return "Extremely Heavy"
        elif rate_mm_hr >= 30:
            return "Heavy"
        elif rate_mm_hr >= 10:
            return "Moderate"
        else:
            return "Light"

    @classmethod
    def fetch_live_weather(cls, lat: float, lng: float) -> Dict[str, Any]:
        """
        Queries Open-Meteo Forecast API for current and hourly weather parameters.
        Returns normalized dictionary with real weather telemetry.
        """
        try:
            url = (
                f"{cls.OPEN_METEO_BASE_URL}?latitude={lat}&longitude={lng}"
                "&current=temperature_2m,relative_humidity_2m,precipitation,rain,showers,weather_code,wind_speed_10m"
                "&hourly=temperature_2m,relative_humidity_2m,precipitation,rain,showers,soil_moisture_0_to_1cm,soil_moisture_1_to_3cm,soil_moisture_3_to_9cm,wind_speed_10m,weather_code"
                "&past_hours=24&forecast_days=2&timezone=auto"
            )
            req = urllib.request.Request(url, headers={"User-Agent": "AapdaSetu/1.0"})
            with urllib.request.urlopen(req, timeout=8) as response:
                data = json.loads(response.read().decode("utf-8"))
                
                current = data.get("current", {})
                hourly = data.get("hourly", {})
                times = hourly.get("time", [])
                
                curr_time = current.get("time", "")
                curr_prefix = curr_time[:13]
                curr_idx = -1
                for i, t in enumerate(times):
                    if t.startswith(curr_prefix):
                        curr_idx = i
                        break
                if curr_idx < 0:
                    curr_idx = 0
                    
                # Read current rain and precipitation (Do NOT convert null to 0)
                raw_precip = current.get("precipitation")
                raw_rain = current.get("rain")
                raw_showers = current.get("showers")
                
                rain_val = float(raw_rain) if raw_rain is not None else (float(hourly.get("rain", [None])[curr_idx]) if curr_idx < len(hourly.get("rain", [])) and hourly.get("rain", [])[curr_idx] is not None else None)
                precip_val = float(raw_precip) if raw_precip is not None else (float(hourly.get("precipitation", [None])[curr_idx]) if curr_idx < len(hourly.get("precipitation", [])) and hourly.get("precipitation", [])[curr_idx] is not None else (rain_val if rain_val is not None else None))
                
                # Helper for slicing & summing hourly metrics safely
                def sum_hourly(arr_key, start_offset, end_offset):
                    arr = hourly.get(arr_key, [])
                    s = max(0, curr_idx + start_offset)
                    e = min(len(arr), curr_idx + end_offset)
                    if s >= e:
                        return None
                    valid = [float(v) for v in arr[s:e] if v is not None]
                    return round(sum(valid), 1) if valid else None

                # Recent hourly accumulations
                accum_1h_rain = sum_hourly("rain", 0, 1) or rain_val
                accum_3h_rain = sum_hourly("rain", -2, 1)
                accum_6h_rain = sum_hourly("rain", -5, 1)
                accum_24h_rain = sum_hourly("rain", -23, 1)

                accum_3h_precip = sum_hourly("precipitation", -2, 1)
                accum_24h_precip = sum_hourly("precipitation", -23, 1)

                # Next 24h forecasts
                forecast_24h_rain = sum_hourly("rain", 0, 24)
                forecast_24h_precip = sum_hourly("precipitation", 0, 24)

                # Soil moisture
                sm_list = hourly.get("soil_moisture_0_to_1cm", [])
                sm_val = float(sm_list[curr_idx]) if sm_list and curr_idx < len(sm_list) and sm_list[curr_idx] is not None else 0.30
                soil_sat_pct = min(100.0, max(0.0, (sm_val / 0.48) * 100.0))
                
                temp = float(current.get("temperature_2m", 20.0)) if current.get("temperature_2m") is not None else None
                wind = float(current.get("wind_speed_10m", 5.0)) if current.get("wind_speed_10m") is not None else None
                
                return {
                    "success": True,
                    "rainfall_rate": round(rain_val, 1) if rain_val is not None else (round(precip_val, 1) if precip_val is not None else None),
                    "rainfall_display": f"{round(rain_val if rain_val is not None else precip_val, 1)} mm/hr" if (rain_val is not None or precip_val is not None) else "Unavailable",
                    "rain_mm_hr": round(rain_val, 1) if rain_val is not None else None,
                    "precipitation_mm_hr": round(precip_val, 1) if precip_val is not None else None,
                    "accum_1h_rain_mm": accum_1h_rain,
                    "accum_3h_rain_mm": accum_3h_rain,
                    "accum_6h_rain_mm": accum_6h_rain,
                    "accum_24h_rain_mm": accum_24h_rain,
                    "accum_24h_precipitation_mm": accum_24h_precip,
                    "forecast_24h_rain_mm": forecast_24h_rain,
                    "forecast_24h_precipitation_mm": forecast_24h_precip,
                    "rainfall_mm": accum_24h_rain or forecast_24h_precip,
                    "rainfall_intensity": cls.get_intensity_label(rain_val if rain_val is not None else precip_val),
                    "soil_saturation_pct": round(soil_sat_pct, 1),
                    "temperature_c": round(temp, 1) if temp is not None else None,
                    "wind_speed_kmh": round(wind, 1) if wind is not None else None,
                    "is_live_data": True,
                    "request_url": url,
                    "matched_hourly_index": curr_idx,
                    "matched_hourly_time": times[curr_idx] if curr_idx < len(times) else None
                }
        except Exception as e:
            print(f"[WeatherService] Open-Meteo fetch error: {e}")
            return {
                "success": False,
                "error": "Live weather data temporarily unavailable.",
                "is_live_data": False
            }

    @staticmethod
    def simulate_heavy_downpour(env_data):
        env_data.rainfall_rate = 145.0
        env_data.rainfall_mm += 115.0
        env_data.rainfall_intensity = "Cloudburst / Torrential"
        env_data.rainfall_forecast_trend = "Peaking"
        return env_data

File: frontend/src/services/weatherService.js
// OPEN-METEO API
// No API key is required for this prototype.
// Weather data is fetched using latitude and longitude.

/**
 * Weather & Precipitation Telemetry Service
 * Connects directly to Open-Meteo Forecast API (https://api.open-meteo.com/v1/forecast)
 * 
 * Accurately retrieves and parses:
 * - current.precipitation, current.rain, current.showers
 * - hourly.precipitation, hourly.rain, hourly.showers (matched to current hour, not index 0)
 * - timezone=auto
 * - Preserves null/missing values without fabricating 0s
 */

const OPEN_METEO_BASE_URL = 'https://api.open-meteo.com/v1/forecast';

// Client-side cache to avoid excessive API requests
const weatherCache = new Map();
const CACHE_TTL_MS = 5 * 60 * 1000; // 5 minutes cache

export const weatherService = {
  /**
   * Fetch live weather and hourly forecast from Open-Meteo
   * @param {number} latitude 
   * @param {number} longitude 
   * @param {boolean} forceRefresh 
   * @returns {Promise<{success: boolean, data?: object, error?: string, request_url?: string}>}
   */
  async fetchLiveWeather(latitude, longitude, forceRefresh = false) {
    const lat = Number(latitude);
    const lng = Number(longitude);

    if (isNaN(lat) || isNaN(lng)) {
      return {
        success: false,
        error: 'Invalid coordinates provided.'
      };
    }

    const cacheKey = `${lat.toFixed(4)},${lng.toFixed(4)}`;
    const cached = weatherCache.get(cacheKey);
    const now = Date.now();

    if (!forceRefresh && cached && (now - cached.timestamp < CACHE_TTL_MS)) {
      return {
        success: true,
        data: cached.data,
        fromCache: true,
        request_url: cached.request_url
      };
    }

    try {
      // Build Open-Meteo Forecast API query with requested current and hourly parameters + timezone=auto
      const url = new URL(OPEN_METEO_BASE_URL);
      url.searchParams.set('latitude', lat.toString());
      url.searchParams.set('longitude', lng.toString());
      url.searchParams.set('current', [
        'temperature_2m',
        'relative_humidity_2m',
        'precipitation',
        'rain',
        'showers',
        'weather_code',
        'wind_speed_10m'
      ].join(','));
      url.searchParams.set('hourly', [
        'temperature_2m',
        'relative_humidity_2m',
        'precipitation',
        'rain',
        'showers',
        'soil_moisture_0_to_1cm',
        'soil_moisture_1_to_3cm',
        'soil_moisture_3_to_9cm',
        'wind_speed_10m',
        'weather_code'
      ].join(','));
      url.searchParams.set('past_hours', '24');
      url.searchParams.set('forecast_days', '2');
      url.searchParams.set('timezone', 'auto');

      const requestUrlStr = url.toString();
      const response = await fetch(requestUrlStr);

      if (!response.ok) {
        console.error(`[weatherService] Open-Meteo returned status ${response.status}`);
        return {
          success: false,
          error: 'Live weather data temporarily unavailable.',
          request_url: requestUrlStr
        };
      }

      const json = await response.json();
      const parsedData = this.parseOpenMeteoPayload(json, lat, lng, requestUrlStr);

      // Cache validated data
      weatherCache.set(cacheKey, {
        timestamp: now,
        data: parsedData,
        request_url: requestUrlStr
      });

      return {
        success: true,
        data: parsedData,
        fromCache: false,
        request_url: requestUrlStr
      };
    } catch (err) {
      console.error('[weatherService] Network / Fetch error querying Open-Meteo:', err);
      return {
        success: false,
        error: 'Live weather data temporarily unavailable.'
      };
    }
  },

  /**
   * Helper to safely sum a range of hourly values (preserving null when all values are missing)
   */
  sumHourlySlice(arr, startIdx, endIdx) {
    if (!Array.isArray(arr) || arr.length === 0) return null;
    const start = Math.max(0, startIdx);
    const end = Math.min(arr.length, endIdx);
    if (start >= end) return null;
    const slice = arr.slice(start, end);
    const valid = slice.filter(v => typeof v === 'number' && !isNaN(v));
    if (valid.length === 0) return null;
    return Math.round(valid.reduce((sum, v) => sum + v, 0) * 10) / 10;
  },

  /**
   * Parse and structure raw Open-Meteo API payload into AapdaSetu-compatible telemetry
   */
  parseOpenMeteoPayload(json, lat, lng, requestUrl = '') {
    const current = json.current || {};
    const hourly = json.hourly || {};
    const times = hourly.time || [];

    // 1. Precise Current Hour Index Matching (Matched to current local ISO timestamp)
    const currentTimeStr = current.time || '';
    const currentHourPrefix = currentTimeStr.length >= 13 ? currentTimeStr.slice(0, 13) : '';
    let currentIdx = times.findIndex(t => t.startsWith(currentHourPrefix));

    if (currentIdx < 0 && currentTimeStr) {
      const targetTs = new Date(currentTimeStr).getTime();
      if (!isNaN(targetTs)) {
        let minDiff = Infinity;
        times.forEach((t, i) => {
          const diff = Math.abs(new Date(t).getTime() - targetTs);
          if (diff < minDiff) {
            minDiff = diff;
            currentIdx = i;
          }
        });
      }
    }
    if (currentIdx < 0) currentIdx = 0;

    // 2. Verified Current Rain & Precipitation Extraction (Do NOT convert null/missing to 0)
    const currentRain = typeof current.rain === 'number'
      ? current.rain
      : (currentIdx >= 0 && typeof hourly.rain?.[currentIdx] === 'number' ? hourly.rain[currentIdx] : null);

    const currentPrecip = typeof current.precipitation === 'number'
      ? current.precipitation
      : (currentIdx >= 0 && typeof hourly.precipitation?.[currentIdx] === 'number' ? hourly.precipitation[currentIdx] : null);

    const currentShowers = typeof current.showers === 'number'
      ? current.showers
      : (currentIdx >= 0 && typeof hourly.showers?.[currentIdx] === 'number' ? hourly.showers[currentIdx] : null);

    // Primary current rain rate (mm/hr)
    const rain_mm_hr = currentRain !== null ? Math.round(currentRain * 10) / 10 : null;
    const precipitation_mm_hr = currentPrecip !== null ? Math.round(currentPrecip * 10) / 10 : (rain_mm_hr !== null ? rain_mm_hr : null);
    const showers_mm_hr = currentShowers !== null ? Math.round(currentShowers * 10) / 10 : null;

    // 3. Historical / Recent Rainfall Accumulations (Preceding and including the current hour)
    // - Last 1 hour: current hourly index
    // - Last 3 hours: currentIdx - 2 through currentIdx (3 hours)
    // - Last 6 hours: currentIdx - 5 through currentIdx (6 hours)
    // - Last 24 hours: currentIdx - 23 through currentIdx (24 hours)
    const accum1hRain = this.sumHourlySlice(hourly.rain, currentIdx, currentIdx + 1) ?? rain_mm_hr;
    const accum3hRain = this.sumHourlySlice(hourly.rain, currentIdx - 2, currentIdx + 1);
    const accum6hRain = this.sumHourlySlice(hourly.rain, currentIdx - 5, currentIdx + 1);
    const accum24hRain = this.sumHourlySlice(hourly.rain, currentIdx - 23, currentIdx + 1);

    const accum1hPrecip = this.sumHourlySlice(hourly.precipitation, currentIdx, currentIdx + 1) ?? precipitation_mm_hr;
    const accum3hPrecip = this.sumHourlySlice(hourly.precipitation, currentIdx - 2, currentIdx + 1);
    const accum6hPrecip = this.sumHourlySlice(hourly.precipitation, currentIdx - 5, currentIdx + 1);
    const accum24hPrecip = this.sumHourlySlice(hourly.precipitation, currentIdx - 23, currentIdx + 1);

    // 4. Next 24 Hours Cumulative Forecast (Future from currentIdx)
    const forecast24hRain = this.sumHourlySlice(hourly.rain, currentIdx, currentIdx + 24);
    const forecast24hPrecip = this.sumHourlySlice(hourly.precipitation, currentIdx, currentIdx + 24);

    // 5. Console Debugging Telemetry
    console.log('🌧️ [AapdaSetu Weather Telemetry]');
    console.log('  📍 Selected latitude:', lat);
    console.log('  📍 Selected longitude:', lng);
    console.log('  💧 Current rain:', currentRain !== null ? `${currentRain} mm/hr` : 'Unavailable');
    console.log('  💧 Current precipitation:', currentPrecip !== null ? `${currentPrecip} mm/hr` : 'Unavailable');
    console.log('  📊 Hourly rain values:', hourly.rain ? hourly.rain.slice(Math.max(0, currentIdx - 23), currentIdx + 24) : 'Unavailable');
    console.log('  ⏱️ Last 1h rainfall:', accum1hRain !== null ? `${accum1hRain} mm` : 'Unavailable');
    console.log('  ⏱️ Last 3h rainfall:', accum3hRain !== null ? `${accum3hRain} mm` : 'Unavailable');
    console.log('  ⏱️ Last 6h rainfall:', accum6hRain !== null ? `${accum6hRain} mm` : 'Unavailable');
    console.log('  ⏱️ Last 24h rainfall (accumulated):', accum24hRain !== null ? `${accum24hRain} mm` : 'Unavailable');
    console.log('  🔮 Next 24h rainfall (forecast):', forecast24hRain !== null ? `${forecast24hRain} mm` : 'Unavailable');

    // 6. Temperature, Wind, Humidity
    const temperature = typeof current.temperature_2m === 'number'
      ? current.temperature_2m
      : (currentIdx >= 0 && typeof hourly.temperature_2m?.[currentIdx] === 'number' ? hourly.temperature_2m[currentIdx] : null);

    const windSpeed = typeof current.wind_speed_10m === 'number'
      ? current.wind_speed_10m
      : (currentIdx >= 0 && typeof hourly.wind_speed_10m?.[currentIdx] === 'number' ? hourly.wind_speed_10m[currentIdx] : null);

    const relativeHumidity = typeof current.relative_humidity_2m === 'number'
      ? current.relative_humidity_2m
      : (currentIdx >= 0 && typeof hourly.relative_humidity_2m?.[currentIdx] === 'number' ? hourly.relative_humidity_2m[currentIdx] : null);

    // 7. Soil Moisture from Open-Meteo Land Surface Model
    const sm0_1 = hourly.soil_moisture_0_to_1cm?.[currentIdx] ?? null;
    const sm1_3 = hourly.soil_moisture_1_to_3cm?.[currentIdx] ?? sm0_1;
    const sm3_9 = hourly.soil_moisture_3_to_9cm?.[currentIdx] ?? sm1_3;
    const avgSoilMoistureM3 = sm0_1 !== null ? (sm0_1 + (sm1_3 ?? sm0_1) + (sm3_9 ?? sm1_3 ?? sm0_1)) / 3 : 0.30;
    const soilSaturationPct = Math.min(100, Math.max(0, Math.round((avgSoilMoistureM3 / 0.48) * 100)));

    // 8. Intensity Label
    const effectiveRate = rain_mm_hr !== null ? rain_mm_hr : (precipitation_mm_hr !== null ? precipitation_mm_hr : 0.0);
    let intensityLabel = 'Light / Dry';
    if (effectiveRate >= 100) intensityLabel = 'Cloudburst / Torrential';
    else if (effectiveRate >= 50) intensityLabel = 'Extremely Heavy';
    else if (effectiveRate >= 30) intensityLabel = 'Heavy';
    else if (effectiveRate >= 10) intensityLabel = 'Moderate';
    else if (effectiveRate > 0) intensityLabel = 'Light Rain';
    else if (rain_mm_hr === null && precipitation_mm_hr === null) intensityLabel = 'Unavailable';

    // 9. Forecast Trend
    const next3hSum = (hourly.precipitation || [])
      .slice(currentIdx, currentIdx + 3)
      .reduce((s, v) => s + (typeof v === 'number' ? v : 0), 0);
    const prev3hSum = (hourly.precipitation || [])
      .slice(Math.max(0, currentIdx - 3), currentIdx)
      .reduce((s, v) => s + (typeof v === 'number' ? v : 0), 0);

    let rainfallTrend = 'Stable';
    if (effectiveRate >= 50 || next3hSum >= 70) rainfallTrend = 'Peaking';
    else if (next3hSum > prev3hSum && next3hSum > 5) rainfallTrend = 'Rising Rapidly';
    else if (next3hSum > 2) rainfallTrend = 'Rising';
    else if (prev3hSum > next3hSum && prev3hSum > 0.5) rainfallTrend = 'Falling';

    // 10. Estimated Surface Runoff
    const estimatedRunoffCoeff = Math.min(0.95, (soilSaturationPct / 100) * 0.85 + (effectiveRate > 25 ? 0.15 : 0.05));
    const estimatedSurfaceRunoffMm = (rain_mm_hr !== null || precipitation_mm_hr !== null)
      ? Math.round(effectiveRate * estimatedRunoffCoeff * 10) / 10
      : null;

    // 11. Grounded Risk Factors
    const riskFactors = [];
    if (rain_mm_hr !== null && rain_mm_hr >= 50) {
      riskFactors.push(`Torrential rainfall rate detected (${rain_mm_hr.toFixed(1)} mm/hr)`);
    } else if (rain_mm_hr !== null && rain_mm_hr >= 25) {
      riskFactors.push(`Heavy rainfall rate active (${rain_mm_hr.toFixed(1)} mm/hr)`);
    } else if (rain_mm_hr !== null && rain_mm_hr >= 10) {
      riskFactors.push(`Moderate rainfall active (${rain_mm_hr.toFixed(1)} mm/hr)`);
    }

    if (accum24hRain !== null && accum24hRain >= 75) {
      riskFactors.push(`High 24h accumulated rainfall recorded (${accum24hRain.toFixed(1)} mm)`);
    } else if (accum24hRain !== null && accum24hRain >= 40) {
      riskFactors.push(`Elevated 24h rainfall accumulation (${accum24hRain.toFixed(1)} mm)`);
    }

    if (forecast24hRain !== null && forecast24hRain >= 75) {
      riskFactors.push(`High 24h rainfall forecast expected (${forecast24hRain.toFixed(1)} mm)`);
    } else if (forecast24hRain !== null && forecast24hRain >= 40) {
      riskFactors.push(`Elevated 24h rainfall forecast (${forecast24hRain.toFixed(1)} mm)`);
    }

    if (soilSaturationPct >= 75) {
      riskFactors.push(`High soil moisture saturation (${soilSaturationPct}% pore-water volume)`);
    } else if (soilSaturationPct >= 55) {
      riskFactors.push(`Elevated soil moisture level (${soilSaturationPct}%)`);
    }

    if (estimatedSurfaceRunoffMm !== null && estimatedSurfaceRunoffMm >= 15) {
      riskFactors.push(`Significant surface runoff load (${estimatedSurfaceRunoffMm.toFixed(1)} mm/hr derived)`);
    }

    if (windSpeed !== null && windSpeed >= 45) {
      riskFactors.push(`High wind gusts (${windSpeed.toFixed(1)} km/h)`);
    }

    if (riskFactors.length === 0) {
      riskFactors.push('Atmospheric and precipitation readings within safe baseline limits');
    }

    return {
      source: 'Open-Meteo Live Forecast API',
      latitude: lat,
      longitude: lng,
      elevation: json.elevation || null,
      timezone: json.timezone || 'UTC',
      current_time: currentTimeStr,
      matched_hourly_index: currentIdx,
      matched_hourly_time: times[currentIdx] || null,
      request_url: requestUrl,
      field_source: currentRain !== null ? 'current.rain' : (currentPrecip !== null ? 'current.precipitation' : `hourly.rain[${currentIdx}]`),
      fetched_at: new Date().toISOString(),

      // Raw Current vs Hourly Readings
      raw_current: {
        precipitation: typeof current.precipitation === 'number' ? current.precipitation : null,
        rain: typeof current.rain === 'number' ? current.rain : null,
        showers: typeof current.showers === 'number' ? current.showers : null
      },
      raw_hourly_current: {
        precipitation: currentIdx >= 0 && typeof hourly.precipitation?.[currentIdx] === 'number' ? hourly.precipitation[currentIdx] : null,
        rain: currentIdx >= 0 && typeof hourly.rain?.[currentIdx] === 'number' ? hourly.rain[currentIdx] : null,
        showers: currentIdx >= 0 && typeof hourly.showers?.[currentIdx] === 'number' ? hourly.showers[currentIdx] : null
      },

      // Processed Telemetry (No fake 0s for missing values)
      rain_mm_hr: rain_mm_hr,
      rain_display: rain_mm_hr !== null ? `${rain_mm_hr} mm/hr` : 'Unavailable',
      precipitation_mm_hr: precipitation_mm_hr,
      precipitation_display: precipitation_mm_hr !== null ? `${precipitation_mm_hr} mm/hr` : 'Unavailable',
      showers_mm_hr: showers_mm_hr,

      // Recent Accumulations (Observed Past Periods)
      accum_1h_rain_mm: accum1hRain,
      accum_3h_rain_mm: accum3hRain,
      accum_6h_rain_mm: accum6hRain,
      accum_24h_rain_mm: accum24hRain,

      accum_1h_precipitation_mm: accum1hPrecip,
      accum_3h_precipitation_mm: accum3hPrecip,
      accum_6h_precipitation_mm: accum6hPrecip,
      accum_24h_precipitation_mm: accum24hPrecip,

      // Future Forecast Accumulations (Next 24 Hours)
      forecast_24h_rain_mm: forecast24hRain,
      forecast_24h_precipitation_mm: forecast24hPrecip,

      // Classification & Atmosphere
      rainfall_intensity: intensityLabel,
      rainfall_forecast_trend: rainfallTrend,
      temperature_c: temperature !== null ? Math.round(temperature * 10) / 10 : null,
      wind_speed_kmh: windSpeed !== null ? Math.round(windSpeed * 10) / 10 : null,
      relative_humidity_pct: relativeHumidity !== null ? Math.round(relativeHumidity) : null,

      // Soil moisture from Open-Meteo
      soil_moisture_m3: Math.round(avgSoilMoistureM3 * 1000) / 1000,
      soil_saturation_pct: soilSaturationPct,

      // Derived Runoff
      surface_runoff: {
        value_mm_hr: estimatedSurfaceRunoffMm,
        is_derived: true,
        derivation_basis: 'Calculated from Open-Meteo precipitation rate & soil moisture saturation'
      },

      // Hourly preview (next 12 hours)
      hourly_forecast: times.slice(currentIdx, currentIdx + 12).map((timeStr, idx) => {
        const offsetIdx = currentIdx + idx;
        return {
          time: timeStr,
          temperature: hourly.temperature_2m?.[offsetIdx] ?? null,
          precipitation: hourly.precipitation?.[offsetIdx] ?? null,
          rain: hourly.rain?.[offsetIdx] ?? null,
          showers: hourly.showers?.[offsetIdx] ?? null,
          windSpeed: hourly.wind_speed_10m?.[offsetIdx] ?? null,
          soilMoisture: hourly.soil_moisture_0_to_1cm?.[offsetIdx] ?? null
        };
      }),

      // Grounded risk factors supported by API values
      risk_factors: riskFactors
    };
  }
};
